From 809a43c50058ef245e7b14e990ed85640e4bca8d Mon Sep 17 00:00:00 2001
From: Hyun Kwon <hyun.kwon@xilinx.com>
Date: Thu, 16 Jul 2020 12:48:27 -0700
Subject: [PATCH 1447/1851] sound: xilinx: pcm: Set the platform device name
 correctly

commit b9ebe4b677fc90e294e0f148707dce895a01dfe8 from
https://github.com/Xilinx/linux-xlnx.git

The driver changes the name of device, and that results in freeing
the original name. But the original name is shared between
the device and its platform device object, and the platform device
object ends up pointing to invalid buffer. This patch updates
the platform name accordingly.

Trimmed down and modified error messages:

[] BUG: KASAN: use-after-free in strcmp+0x28/0x70
..
[] Call trace:
[] dump_backtrace+0x0/0x1e8
[] show_stack+0x14/0x20
[] dump_stack+0xd4/0x108
[] print_address_description.isra.0+0x68/0x324
[] __kasan_report+0x144/0x198
[] kasan_report+0xc/0x18
[] __asan_load1+0x5c/0x68
[] strcmp+0x28/0x70
[] platform_match+0x104/0x118
[] __driver_attach+0x40/0x108
...

[] Allocated by task 38:
[] save_stack+0x24/0xb0
[] __kasan_kmalloc.isra.0+0xc0/0xe0
[] kasan_kmalloc+0xc/0x18
[] __kmalloc_track_caller+0x104/0x1d8
[] kvasprintf+0xbc/0x158
[] kvasprintf_const+0xc4/0x170
[] kobject_set_name_vargs+0x50/0xe8
[] dev_set_name+0xa0/0xd0
[] of_device_make_bus_id+0x194/0x1a8
[] of_device_alloc+0x208/0x228
[] of_platform_device_create_pdata+0xb8/0x168
[] of_platform_bus_create+0x244/0x4e0
[] of_platform_populate+0x50/0xe8
[] zynqmp_dpsub_probe+0xf4/0x178
...

[] Freed by task 38:
[] save_stack+0x24/0xb0
[] __kasan_slab_free+0xf4/0x158
[] kasan_slab_free+0x10/0x18
[] kfree+0x84/0xf8
[] kfree_const+0x1c/0x38
[] kobject_set_name_vargs+0xa0/0xe8
[] dev_set_name+0xa0/0xd0
[] xilinx_dp_pcm_probe+0x40/0x80
...

Signed-off-by: Hyun Kwon <hyun.kwon@xilinx.com>
Tested-by: Rohit Visavalia <rohit.visavalia@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 sound/soc/xilinx/xilinx-dp-pcm.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/sound/soc/xilinx/xilinx-dp-pcm.c b/sound/soc/xilinx/xilinx-dp-pcm.c
index fa8abe788cf7..36c5afb50f4b 100644
--- a/sound/soc/xilinx/xilinx-dp-pcm.c
+++ b/sound/soc/xilinx/xilinx-dp-pcm.c
@@ -47,6 +47,7 @@ static int xilinx_dp_pcm_probe(struct platform_device *pdev)
 	int ret;
 
 	dev_set_name(&pdev->dev, pdev->dev.of_node->name);
+	pdev->name = dev_name(&pdev->dev);
 	ret = devm_snd_dmaengine_pcm_register(&pdev->dev,
 					      &xilinx_dmaengine_pcm_config, 0);
 	if (ret)
-- 
2.31.1

